app/src/led_indicators.c

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2022 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+#include <zmk/ble.h>
+#include <zmk/endpoints.h>
+#include <zmk/led_indicators.h>
+#include <zmk/events/led_indicator_changed.h>
+#include <zmk/events/endpoint_selection_changed.h>
+#include <zmk/split/bluetooth/central.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#define NUM_USB_PROFILES (COND_CODE_1(IS_ENABLED(CONFIG_ZMK_USB), (1), (0)))
+#define NUM_BLE_PROFILES (COND_CODE_1(IS_ENABLED(CONFIG_ZMK_BLE), (ZMK_BLE_PROFILE_COUNT), (0)))
+#define NUM_PROFILES (NUM_USB_PROFILES + NUM_BLE_PROFILES)
+
+static zmk_leds_flags_t led_flags[NUM_PROFILES];
+
+static size_t profile_index(enum zmk_endpoint endpoint, uint8_t profile) {
+    switch (endpoint) {
+    case ZMK_ENDPOINT_USB:
+        return 0;
+    case ZMK_ENDPOINT_BLE:
+        return NUM_USB_PROFILES + profile;
+    }
+
+    CODE_UNREACHABLE;
+}
+
+zmk_leds_flags_t zmk_leds_get_current_flags() {
+    enum zmk_endpoint endpoint = zmk_endpoints_selected();
+    uint8_t profile = 0;
+
+#if IS_ENABLED(CONFIG_ZMK_BLE)
+    if (endpoint == ZMK_ENDPOINT_BLE) {
+        profile = zmk_ble_active_profile_index();
+    }
+#endif
+
+    return zmk_leds_get_flags(endpoint, profile);
+}
+
+zmk_leds_flags_t zmk_leds_get_flags(enum zmk_endpoint endpoint, uint8_t profile) {
+    size_t index = profile_index(endpoint, profile);
+    return led_flags[index];
+}
+
+static void raise_led_changed_event(struct k_work *_work) {
+    zmk_leds_flags_t flags = zmk_leds_get_current_flags();
+
+    ZMK_EVENT_RAISE(new_zmk_led_changed((struct zmk_led_changed){.leds = flags}));
+
+#if IS_ENABLED(CONFIG_ZMK_SPLIT) && IS_ENABLED(CONFIG_ZMK_BLE)
+    zmk_split_bt_update_leds(flags);
+#endif
+}
+
+static K_WORK_DEFINE(led_changed_work, raise_led_changed_event);
+
+void zmk_leds_update_flags(zmk_leds_flags_t leds, enum zmk_endpoint endpoint, uint8_t profile) {
+    size_t index = profile_index(endpoint, profile);
+    led_flags[index] = leds;
+
+    k_work_submit(&led_changed_work);
```

Comment:
I can add a mutex or an atomic variable but I don't see why that is necessary. In this function we never read the value, we just write it and writing operations are already atomic.

Reply:
I was more concerned about the other functions that do read the value.

Also, do you think it matters whether two changes here in quick succession result in a single notification of the final state or separate notifications for each change? If we want the latter, we really need to use a message queue (or similar) so the main thread can process each individual change.

Reply:
Okay. This is fine then, though I would still be concerned about someone later adding code which reads the variable and not realizing that it gets written on another thread. Adding a comment or making it an atomic would at least be a clue that more care needs to be taken because multiple threads are involved.

Reply:
I added a comment.